xdsl/parser.py

```diff
@@ -1,1335 +1,1942 @@
-            return True
-        return None
+    def _find_token_end(self, start: int | None = None) -> int:
+        """
+        Find the point (optionally starting from start) where the token ends
+        """
+        i = self.next_pos() if start is None else start
+        # search for literal breaks
+        for part in self.break_on:
+            if self.input.content.startswith(part, i):
+                return i + len(part)
+        # otherwise return the start of the next break
+        return min(
+            filter(
+                lambda x: x >= 0,
+                (self.input.content.find(part, i) for part in self.break_on),
+            ))
+
+    def next_pos(self, i: int | None = None) -> int:
+        """
+        Find the next starting position (optionally starting from i)
 
-    def parse_optional_char(self,
-                            char: str,
-                            skip_white_space: bool = True) -> bool | None:
-        assert len(char) == 1
-        if skip_white_space:
-            self.skip_white_space()
-        if self._pos is None:
-            return None
-        if self._pos.get_char() == char:
-            self._pos = self._pos.next_char_pos()
-            return True
-        return None
+        This will skip line comments!
+        """
+        i = self.pos if i is None else i
+        # skip whitespaces
+        while self.input.at(i).isspace():
+            i += 1
 
-    def parse_char(self, char: str, skip_white_space: bool = True) -> bool:
-        assert (len(char) == 1)
-        res = self.parse_optional_char(char, skip_white_space=skip_white_space)
-        if res is None:
-            raise ParserError(self._pos, f"'{char}' expected")
-        return True
-
-    def parse_string(self,
-                     contents: str,
-                     skip_white_space: bool = True) -> bool:
-        if skip_white_space:
-            self.skip_white_space()
-        chars = self.get_char(len(contents))
-        if chars == contents:
-            assert self._pos
-            self._pos = self._pos.next_char_pos(len(contents))
-            return True
-        raise ParserError(self._pos, f"'{contents}' expected")
-
-    def parse_optional_string(self,
-                              contents: str,
-                              skip_white_space: bool = True) -> bool | None:
-        if skip_white_space:
-            self.skip_white_space()
-        chars = self.get_char(len(contents))
-        if chars == contents:
-            assert self._pos is not None
-            self._pos = self._pos.next_char_pos(len(contents))
+        # skip comments as well
+        if self.input.content.startswith("//", i):
+            i = self.input.content.find("\n", i) + 1
+            return self.next_pos(i)
+
+        return i
+
+    def is_eof(self):
+        """
+        Check if the end of the input was reached.
+        """
+        try:
+            self.next_pos()
+            return False
```

Comment:
Could you not implement this by looking at the position and comparing it with the length of the input?

Reply:
That would break for whitespace or comments after the last token